### PR TITLE
Improve c2cpg performance.

### PR DESCRIPTION
**Pull request by ml86**
- Moved ast creation out of global lock in order to avoid threads waiting for the
  lock the majority of the time.
- Improve column calculation algorithm.

    The old version managed to be slow enough to be the number one hot spot
    during profiling of c2cpg. The new version makes this fast so that it does
    not show up during performance measuring anymore.

    Column number now start at one instead of zero like the line numbers and
    the new implemention obsoleted a bug in the old column end calculation
    which led to complete bogus numbers.

### CODE REVIEW

File: joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala [hunk 1]
```diff
@@ -58,9 +59,25 @@ trait AstCreatorHelper {
     }
   }
 
-  private def fileLines(node: IASTNode): Seq[Int] = {
+  private def fileOffsetTable(node: IASTNode): Array[Int] = {
     val f = fileName(node)
-    global.file2LinesCache.computeIfAbsent(f, _ => IOUtils.readLineLengthsInFile(Paths.get(f)))
+    global.file2OffsetTable.computeIfAbsent(f, _ => genFileOffsetTable(Paths.get(f)))
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala:64]: Then we could remove `IOUtils.readLineLengthsInFile` as its not needed any longer.

File: joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala [hunk 2]
```diff
@@ -58,9 +59,25 @@ trait AstCreatorHelper {
...
+    asString = asString.replaceAll("\r\n", "\n")
+    asString = asString.replaceAll("\r", "\n")
+    val asCharArray = asString.toCharArray
+    val offsets     = mutable.ArrayBuffer.empty[Int]
+
+    for (i <- Range(0, asCharArray.length)) {
+      if (asCharArray(i) == '\n') {
+        offsets.append(i + 1)
+      }
+    }
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala:79]: Is this truly platform-independent? The IOUtils file reading from cpg does quite some stuff to handle all kinds of problems with charset encoding and unmappable characters.

**Review by max-leuthaeuser — APPROVED** [review]
Just some nits. Thanks a lot! Should bring in some drastic performance improvements.

File: joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala (hunk 2 above)
Comment (ml86) [reply joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/AstCreatorHelper.scala]: I changed the file reading to use IOUtils.

**Pull request merged**